# 1. Downloading and assembling microbial sequence data

Welcome to the **Downloading and assembling microbial sequence data** supplementary materials. This document is written as a [Jupyter](https://jupyter.org/) notebook, which means that it contains both written documentation as well as runnable code in the same document.

There are two methods you can use to run the code here: (1) either copy-paste the bash commands to a separate terminal, or (2) you can run directly in Juptyer.

## 1.1. Copy-paste commands to a separate terminal

In this case, for any commands in a light-gray background, simply copy/paste them to an appropriate terminal environment.

For example, in the below case copy/paste `conda create -y -n sra-tools sra-tools` to install **sra-tools** using conda.

![copy-paste-terminal.png](../images/copy-paste-terminal.png)

## 1.2. Running from within Jupyter

The easiest way to run the code is to click the **Play** button in the Jupyter notebook as shown below:

![run-jupyter.png](../images/run-jupyter.png)

This will run a single cell (block of code or text) and advance to the next cell. A cell that's running will be marked with an asterisk `*`:

![jupyter-cell-run.png](../images/jupyter-cell-run.png)

If you, instead, which to run everything in this notebook all at once you can find the **Kerenel** option in the menu at the top and select **Restart Kerenel and Run All Cells..**, which will run everything in this notebook all at once.

![restart-kernel-jupyter.png](../images/restart-kernel-jupyter.png)

# 2. Download genomes from NCBI

In this section we will go over how to download genomes from [NCBI's Sequence Read Archive](https://www.ncbi.nlm.nih.gov/sra/) by using the [SRA Tools](https://github.com/ncbi/sra-tools) command-line software provided by NCBI.

## 2.1. Configure `conda`

We fist need to configure conda to be able to find the appropriate tools to install as per the [Bioconda instructions](https://bioconda.github.io/user/install.html#set-up-channels). This only needs to be done once.

In [1]:
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge

## 2.2. Install `sra-tools` using conda

Now we can install `sra-tools`. The below command will create a new conda environment `sra-tools` and install the package `sra-tools` in this environment.

The `-y` option here means to automatically answer `yes` to any prompts for input by conda. This is important when running using Jupyter (but can be left out if you are running via the command-line).

In [2]:
conda create -y -n sra-tools sra-tools

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/sra-tools

  added / updated specs:
    - sra-tools


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0
  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0
  curl               conda-forge/linux-64::curl-7.81.0-h2574ce0_0
  hdf5               conda-forge/linux-64::hdf5-1.10.6-nompi_h6a2412b_1114
  icu                conda-forge/linux-64::icu-69.1-h9c3ff4c_0
  krb5               conda-forge/linux-64::krb5-1.19.2-hcc1bbae_3
  libcurl            conda-forge/linux-64::libcurl-7.81.0-h2574ce0_0
  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2
  libev              conda-forge/linux-64::libev-4.33-h516909a_1
  libgcc-ng      

## 2.3. Verify `sra-tools` is working

Let's verify that the tools found in the package `sra-tools` are working properly. We can do this by running one of the installed commands, `prefetch` like so:

In [3]:
conda run -n sra-tools prefetch --help


Usage: prefetch [ options ] [ accessions(s)... ]

Parameters:

  accessions(s)                    list of accessions to process


Options:

  -T|--type <file-type>            Specify file type to download. Default: sra
  -N|--min-size <size>             Minimum file size to download in KB
                                     (inclusive).
  -X|--max-size <size>             Maximum file size to download in KB
                                     (exclusive). Default: 20G
  -f|--force <no|yes|all|ALL>      Force object download - one of: no, yes,
                                     all, ALL. no [default]: skip download if
                                     the object if found and complete; yes:
                                     download it even if it is found and is
                                     complete; all: ignore lock files (stale
                                     locks or it is being downloaded by
                                     another process - use at your own

Notice that I ran by first specifying `conda run -n sra-tools [...]`. The `conda run` part lets you specify a particular environment to activate first before trying to run a tool. It's useful in cases where you are installing each tool in a separate environment (as we are doing here). You can also use `conda activate sra-tools` first and then run `prefetch --help` (but this will not work if running in Jupyter).

## 2.4. Prefetch genomes (`prefetch`)

The [prefetch](https://github.com/ncbi/sra-tools/wiki/08.-prefetch-and-fasterq-dump) command that is part of the `sra-tools` package can be used to download and store sequence read data from NCBI's Sequence Read Archive. You can run it like so:

In [4]:
conda run -n sra-tools prefetch SRR3028792


2022-01-24T16:23:31 prefetch.2.11.0: 1) 'SRR3028792' is found locally
2022-01-24T16:23:31 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies




You can even pass multiple accession numbers to `prefetch` like so:

In [5]:
conda run -n sra-tools prefetch SRR3028792 SRR3028793


2022-01-24T16:23:32 prefetch.2.11.0: 1) 'SRR3028792' is found locally
2022-01-24T16:23:32 prefetch.2.11.0: 'SRR3028792' has 0 unresolved dependencies

2022-01-24T16:23:33 prefetch.2.11.0: 2) 'SRR3028793' is found locally
2022-01-24T16:23:33 prefetch.2.11.0: 'SRR3028793' has 0 unresolved dependencies




The `prefetch` command downloads the data from NCBI and stores on your computer in a special directory by default. You can see which directory using the `srapath` command:

In [6]:
conda run -n sra-tools srapath SRR3028792

/home/CSCScience.ca/apetkau/ncbi/public/sra/SRR3028792.sra




This tells us that there is a file `SRR3028792.sra` that has been downloaded to the path above which contains the sequence data.

For how to know what identifier to use for `prefetch`, this can be found on the NCBI website.

For example for `SRR3028792`, you can find this identifier (and the associated biological sample) at https://www.ncbi.nlm.nih.gov/sra/?term=SRR3028792

![genome-sra-identifier.png](../images/genome-sra-identifier.png)

## 2.5. Convert to fastq (`fasterq-dump`)

Once you've prefetched the sequence data you can convert to the **fastq** format using `fasterq-dump` (which is named based on the older `fastq-dump` tool). We need to convert from the `.sra` format to the `.fastq` format as most tools work with **fastq** files (instead of **sra** files, which is specific to NCBI's sequence read archive).

In [7]:
conda run -n sra-tools fasterq-dump SRR3028792

spots read      : 824,262
reads read      : 1,648,524
reads written   : 1,648,524




This will write two `*.fastq` files do your current directory which contain the forward (`_1.fastq`) and reverse (`_2.fastq`) reads.

We can use the `ls` command to **l**i**s**t the files and show additional information (such as file size).

In [8]:
ls -lh *.fastq

-rw-r--r-- 1 apetkau grp_apetkau 396M Jan 24 10:23 SRR3028792_1.fastq
-rw-r--r-- 1 apetkau grp_apetkau 398M Jan 24 10:23 SRR3028792_2.fastq



The `396M` part shows us that `SRR3028792_1.fastq` is **396 MB**. The `-lh` part is interpreted as follows: `l` is for **long list**, which means extra information (such as file size) is printed, `h` is for **human readable** which means that file sizes are converted into easier to read units (like MB, GB, etc) instead of being left as bytes. 

We can compress these files to save on storage space using the `gzip` command.

*Tip: To speed this up, you can use `pigz` (parallel gzip) instead of `gzip`, which will take advantage of multiple processing cores on your machine. Just run `pigz *.fastq`. If you don't have `pigz` installed it can be installed with `conda install pigz`.*

In [9]:
gzip *.fastq

In [10]:
ls -lth *.fastq.gz

-rw-r--r-- 1 apetkau grp_apetkau 110M Jan 24 10:23 SRR3028792_1.fastq.gz
-rw-r--r-- 1 apetkau grp_apetkau 138M Jan 24 10:23 SRR3028792_2.fastq.gz



Now the files end in `.fastq.gz` instead of `.fastq` and are much smaller (`SRR3028792_1.fastq.gz` is **110 MB** compared to **396 MB** for the uncompressed file).

We can also look at the file contents using the `cat` (or `zcat` for files ending in `.gz`) commands along with `head`.

In [11]:
zcat SRR3028792_1.fastq.gz | head

@SRR3028792.1 1 length=90
CTTCATAATCAGGCGATAAATGCCCACCACTTTAGGCTTTCTGGCGCGGATAGCCTCCCCAATAAAATCTTTACGCGTACGCTTTGCTTC
+SRR3028792.1 1 length=90
AABC-C--CE,-,C++@,,,<,CCC,BB,:CCE,,,;CCCCC,,6+,+++8,,:,99,,8,,:,,,,<CCEE9C,@+8+8++88C,9CCC
@SRR3028792.2 2 length=271
ATACACGCCAGCGCGAATATTTATCGCTAGCATCAGGAAAACCAACGAAACAGCCCTCGTTGATTTGCCCCCTGAACAGTGGGCTTTTTTCGCGTACAGATGGCTCCGGCATTAGACCTGTGACATCCCGATACTATTCACATTCCGCAGCTCATACTTCACTTCTGCCCGTAATACGCCTTCGGACCATGCTTACGCATGACGTGTTTATTCATCAGATCATCGTCGATGGGATTCAGCGCCGGGTTAATGCCGCGCGCAATCCACGCCA
+SRR3028792.2 2 length=271
-86A<F-@F7+C+C@77C,CFF9EE8FE7+C,CE,,,,,,,;@,,8,,,,+,,9:,9,+::,,CEE,CEEDEF7,,:,,,,,,:CEFFEB=+8+++++,,:,,49A?++++8B,,,:??,:,,:,CEFF7+8++:,:A@,>,DDDD,+++88@,@,@CCC,@@B>@,>@D*@**6,5*4*6>?***>:*<,5=>,>*;5=*,,*2*<:C+<CF7A?+++3**;:*9C))7+***29F**9)7<5))8>)*9*8;5;)7)7((-57)5(4:(
@SRR3028792.3 3 length=181
TGTCGAGATGGTGGCTGAACTGCATCGGGTCTATAGCGCCCAGACTCAGGCGTATGTCCTGTTTAATGAAGAGAGTGCACTTTCGCAAGCGCTTCTGTTGCAACGTCATGGGGAAGAAGATTTTCTTGCCTTTCTCCGGGC

The command `zcat` uncompresses any `*.gz` files and then prints out the contents. Since these are large files we likely only want to look at the first few lines in the file. The command `head` will print out the first 10 lines by default (and ignore all the rest). The `|` character tells bash that we want to take the output of `zcat SRR3028792_1.fastq.gz` (which will print all the lines of a file) and forward it into the `head` command (which will print out the first 10 lines of data and ignore the rest).

The end result is we see the first 10 lines of the `SRR3028792_1.fastq.gz` file.

We can do the same thing for the reverse reads.

In [12]:
zcat SRR3028792_2.fastq.gz | head

@SRR3028792.1 1 length=300
CTCTCAAACCTTCCTCGTTACTTTTTTCTTTCCGCCTCTCTCCTCCCCCCCCCGCCTCCCCTCTTTTCCCTTTCTCTCCTGCTTCTGCCCCTCTCTCTTCTCCTCCTCTGCCTCTTCCTCCTCCCCCTTACTTCTTCCTCTCTTTCCTCCCCCTCTCCTTCCTCCTCCCCCCCTCTCCCTCCCCCCTTCCCTCCCCTCTCCCTCCCCCCCCTCCCCCCCCCTCCCTCCCTCCTCCCCCCTCTCTCTCCCCTCTCTCCCTCCCCCCTCCCCCCCCTCTCCTCCCCCCCCCCCCTCCCCCCC
+SRR3028792.1 1 length=300
--,-8,-,;:,;,;,;,:,,,,,6;;@,<@,;;,,8,8;,66;;;,88++++++678,,,,6,,6,,,6,9:?,59,:95,,:9,:,,,,9:+995:9:,:,9,4,999,,9,94,99,,8,,,++84,,,,3,8,,,,,63,,3,,33,+++++,6,,,,,,,,,,,+++*******0******))))*)+*))))))0)))))))))(((((((((((,((,((,(((((((((((((,(())))(,((((.))(((((((((((((((((((,((()((((((((,((((,,(((((
@SRR3028792.2 2 length=300
TTGCTTCGATTTCTCTCTCCCTTCACCCGGCGCTGACTCCCCTCTCCTCTTCTCTGATGCCTCCTCACTTCATCCTTCCTCCTCCTCCCCCTTCGTATTACTTTCCTCCGTCCCTTCTGCTCTCCTCCCTCTCCCTCTTTTCTCTCTTTCACATTTCTACTTCCGCCCCCCTCTTTCCCCCCACCCACCCCCCTTTTCCGCGTGCCTCTCCACTCTCGCTTTTTCCTTCCTTCTCCTGTTTCTCCCTCTCACTCTTCCCTCTTGCTTTTTTCTTTCTCTTCTACCCCTCTGCCCCTTCCC
+SRR3028792.2 2 

*Tip: In addition to `head` which prints the first few lines of a file, there is also `tail` which prints the last few lines of a file.*

# 3. Reduce dataset size (`seqtk`) (optional)

For the purpose of this tutorial it will be useful to reduce the number of reads we are trying to process by selecting a random sample of them. This speeds up all the later steps (below) at the expense of possibly leading to a more fragmented assembly.

If you were assembling a genome in a non-tutorial setting it may make sense to use all the sequence read data (and so you can skip this step). But otherwise, to reduce our dataset size please follow the steps below.

## 3.1. Install `seqtk`

We will use the software [seqtk](https://github.com/lh3/seqtk) to select a random sample of reads from the `*.fastq.gz` files. To install this software please run.

In [13]:
conda create -y -n seqtk seqtk

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/seqtk

  added / updated specs:
    - seqtk


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11
  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11
  libzlib            conda-forge/linux-64::libzlib-1.2.11-h36c2ea0_1013
  seqtk              bioconda/linux-64::seqtk-1.3-h5bf99c6_3
  zlib               conda-forge/linux-64::zlib-1.2.11-h36c2ea0_1013


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate seqtk
#
# To deactivate an active environment, use
#
#     $ conda deactivate




In [14]:
conda run -n seqtk seqtk sample

ERROR conda.cli.main_run:execute(33): Subprocess for 'conda run ['seqtk', 'sample']' command failed.  (See above for error)

Usage:   seqtk sample [-2] [-s seed=11] <in.fa> <frac>|<number>

Options: -s INT       RNG seed [11]
         -2           2-pass mode: twice as slow but with much reduced memory





## 3.2. Select a random sample of reads

We will select a random sample of reads to reduce our data to around **10%** of the original size. To do this, please run the following.

In [15]:
conda run --name seqtk seqtk sample -s 11 SRR3028792_1.fastq.gz 0.10 | gzip > SRR3028792_1.sample.fastq.gz
conda run --name seqtk seqtk sample -s 11 SRR3028792_2.fastq.gz 0.10 | gzip > SRR3028792_2.sample.fastq.gz

The way this works is as follows:

* First, keep in mind we have two files `SRR3028792_1.fastq.gz` and `SRR3028792_2.fastq.gz`, each containing one pair of reads for a particular fragment of DNA. We must sample each file separately, but we must make sure that each file was sampled in such a way that the pairs of reads for each DNA fragment are in the same order in the output fastq.gz files. This is achieved by using the same random seed (`-s 11`).
* `-s 11`: The seed used to initate the random number generator. Using an identical number here (it can be any number, we use `11`) means that each time you run the software you will get an identical random sample from the `.fastq.gz` file.
* `SRR3028792_1.fastq.gz` and `SRR3028792_2.fastq.gz`: the two input files.
* `0.10`: The sampling proportion (`0.10` means we will select a random subset of **10%** of the reads).
* `| gzip`: The `|` character forwards the output of the `seqtk` command as input to `gzip`. This has the net result of compressing the subsampled reads.
* `> SRR3028792_1.sample.fastq.gz` and `> SRR3028792_2.sample.fastq.gz`: The `>` character writes the output of the `gzip` command to a file with the given name. This allows us to save the subsampled reads.

We can look at these files with the `ls` command:

In [16]:
ls -lh *.fastq.gz

-rw-r--r-- 1 apetkau grp_apetkau 110M Jan 24 10:23 SRR3028792_1.fastq.gz
-rw-r--r-- 1 apetkau grp_apetkau  11M Jan 24 10:25 SRR3028792_1.sample.fastq.gz
-rw-r--r-- 1 apetkau grp_apetkau 138M Jan 24 10:23 SRR3028792_2.fastq.gz
-rw-r--r-- 1 apetkau grp_apetkau  14M Jan 24 10:25 SRR3028792_2.sample.fastq.gz



Notice from the file sizes that the subsampled reads (e.g., `SRR3028792_1.sample.fastq.gz` is about **10%** the size of the original reads file `SRR3028792_1.fastq.gz`).

We will now use the `SRR3028792_1.sample.fastq.gz` and `SRR3028792_2.sample.fastq.gz` for later steps. To skip subsampling reads just use the `SRR3028792_1.fastq.gz` and `SRR3028792_2.fastq.gz` files instead for later steps.

# 4. Quality filter files (`fastp`)

The software [fastp](https://github.com/OpenGene/fastp) can be used to generate a report of the quality of the sequence reads as well as remove any poor-quality reads which might cause issues for downstream analysis.

## 4.1. Install `fastp` using conda

We can first install `fastp` using conda.

In [17]:
conda create -y -n fastp fastp

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/fastp

  added / updated specs:
    - fastp


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  fastp              bioconda/linux-64::fastp-0.23.2-hd36eab0_1
  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4
  libdeflate         conda-forge/linux-64::libdeflate-1.9-h7f98852_0
  libgcc-ng          conda-forge/linux-64::libgcc-ng-11.2.0-h1d223b6_11
  libgomp            conda-forge/linux-64::libgomp-11.2.0-h1d223b6_11
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-11.2.0-he4da1e4_11


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate fastp
#
# To deactivate an active environment, use
#
#     $ conda deactivate




Now let's take a look at all the different options available for running `fastp`.

In [18]:
conda run -n fastp fastp --help

usage: fastp [options] ... 
options:
  -i, --in1                            read1 input file name (string [=])
  -o, --out1                           read1 output file name (string [=])
  -I, --in2                            read2 input file name (string [=])
  -O, --out2                           read2 output file name (string [=])
      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1. Default is to discard it. (string [=])
      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2. If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be written to this same file. (string [=])
      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched base. (string [=])
      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])
  -m, --merge  

## 4.2. Evaluate quality of sequence reads

To run `fastp` on our set of sequence reads, we can pass in the `*.fastq.gz` files to the command and specify the output file names.

In [19]:
conda run -n fastp fastp --in1 SRR3028792_1.sample.fastq.gz --in2 SRR3028792_2.sample.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz --html SRR3028792.html --json SRR3028792.json

# Uncomment below line to run fastp (and later steps) on full dataset
# conda run -n fastp fastp --in1 SRR3028792_1.fastq.gz --in2 SRR3028792_2.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz --html SRR3028792.html --json SRR3028792.json

Read1 before filtering:
total reads: 82644
total bases: 17684154
Q20 bases: 17216930(97.358%)
Q30 bases: 16355846(92.4887%)

Read2 before filtering:
total reads: 82644
total bases: 17782622
Q20 bases: 15329956(86.2075%)
Q30 bases: 13415583(75.4421%)

Read1 after filtering:
total reads: 79043
total bases: 16703489
Q20 bases: 16392251(98.1367%)
Q30 bases: 15698505(93.9834%)

Read2 after filtering:
total reads: 79043
total bases: 16685789
Q20 bases: 14921766(89.428%)
Q30 bases: 13220149(79.23%)

Filtering result:
reads passed filter: 158086
reads failed due to low quality: 7202
reads failed due to too many N: 0
reads failed due to too short: 0
reads with adapter trimmed: 4674
bases trimmed due to adapters: 105179

Duplication rate: 0.337593%

Insert size peak (evaluated by paired-end reads): 35

JSON report: SRR3028792.json
HTML report: SRR3028792.html

fastp --in1 SRR3028792_1.sample.fastq.gz --in2 SRR3028792_2.sample.fastq.gz --out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.

The meaning of the following command-line arguments is as follows:

* `--in1 SRR3028792_1.sample.fastq.gz --in2 SRR3028792_2.sample.fastq.gz`: The first pair of fastq reads (`--in1`) and second pair (`--in2`) to examine.
* `--out1 SRR3028792_1.fp.fastq.gz --out2 SRR3028792_2.fp.fastq.gz`: The output file names for reads that pass the default quality filters, `--out1` for the first pair and `--out2` for the second pair.
* `--html SRR3028792.html`: The name of the HTML report, which can be opened in a web browser.
* `--json SRR3028792.json`: The name of the JSON report (an alternative format to HTML).

Once `fastp` is finished you can open up the `SRR3028792.html` report in any web browser. In Jupyter you can right-click and select **Open in New Browser Tab**:

![jupyter-fastp-report.png](../images/jupyter-fastp-report.png)

*Note: You can find a pre-computed **fastp** report at `example/SRR3028792-fastp.html`.*

# 5. Genome assembly (`skesa`)

Now that we have removed poor-quality reads from our **fastq** files we can move on to genome assembly, which attempts to fit all the smaller sequence reads in the **fastq** files together to produce longer contiguous sequences (contigs). Ideally you would have one contig per chromosome/plasmid but often chromosomes/plasmids will be broken up into smaller contigs, normally at repetitive regions on the genome that are longer than the read length.

## 5.1. Install skesa

We will use [skesa](https://github.com/ncbi/SKESA) to assemble the sequence reads. Skesa is a *de-novo* genome assembler that is designed specifically for microbial genomes. The first step is installing the software using conda.

In [20]:
conda create -y -n skesa skesa

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/skesa

  added / updated specs:
    - skesa


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    numpy-1.22.1               |   py38h6ae9a64_0         6.9 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         6.9 MB

The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  boost              conda-forge/linux-64::boost-1.70.0-py38h9de70de_1
  boost-cpp          conda-forge/linux-64::boost-cpp-1.70.0-h7b93d67_3
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0
  icu      

In [21]:
conda run -n skesa skesa --help


General options:
  -h [ --help ]                 Produce help message
  -v [ --version ]              Print version
  --cores arg (=0)              Number of cores to use (default all) [integer]
  --memory arg (=32)            Memory available (GB, only for sorted counter) 
                                [integer]
  --hash_count                  Use hash counter [flag]
  --estimated_kmers arg (=100)  Estimated number of unique kmers for bloom 
                                filter (millions, only for hash counter) 
                                [integer]
  --skip_bloom_filter           Don't do bloom filter; use --estimated_kmers as
                                the hash table size (only for hash counter) 
                                [flag]

Input/output options : at least one input providing reads for assembly must be specified:
  --reads arg                   Input fasta/fastq file(s) for reads (could be 
                                used multiple times for different ru

## 5.2. Assemble genome

Next we will run `skesa` on our filtered **fastq** files.

In [22]:
conda run -n skesa skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta

skesa --reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz --contigs_out SRR3028792.fasta 

Total mates: 158086 Paired reads: 79043
Reads acquired in  1.401274s wall, 1.390000s user + 0.000000s system = 1.390000s CPU (99.2%)

Kmer len: 19
Raw kmers: 30543730 Memory needed (GB): 0.58644 Memory available (GB): 29.9865 1 cycle(s) will be performed
Distinct kmers: 156
Kmer count in  0.625688s wall, 11.320000s user + 1.580000s system = 12.900000s CPU (2061.7%)
Uniq kmers merging in  0.022581s wall, 0.020000s user + 0.030000s system = 0.050000s CPU (221.4%)
Adapters: 0 Reads before: 158086 Sequence before: 33389278 Reads after: 158086 Sequence after: 33389278 Reads clipped: 0
Adapters clipped in  0.648459s wall, 11.340000s user + 1.610000s system = 12.950000s CPU (1997.0%)

Kmer len: 21
Raw kmers: 30227558 Memory needed (GB): 0.580369 Memory available (GB): 29.9776 1 cycle(s) will be performed
Distinct kmers: 4403385
Kmer count in  0.277599s wall, 11.160000s user + 0.080000s system = 11

The meaning of the command-line arguments is as follows:

* `--reads SRR3028792_1.fp.fastq.gz,SRR3028792_2.fp.fastq.gz`: The sequence reads. For paired-end data each pair of files needs to be separated by a comma `,`.
* `--contigs_out SRR3028792.fasta`: The output file containing the longer contiguous sequences (contigs), in **fasta** format.

## 5.3. Examine output contigs (`seqkit`)

We can examine the output contigs file quickly on the command-line using the [seqkit](https://bioinf.shenwei.me/seqkit/) command-line tool (this is different from the `seqtk` tool we used before).

### 5.3.1. Install `seqkit`

You can install `seqkit` with the below commands.

In [23]:
conda create -y -n seqkit seqkit

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/seqkit

  added / updated specs:
    - seqkit


The following NEW packages will be INSTALLED:

  seqkit             bioconda/linux-64::seqkit-2.1.0-h9ee0642_0


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate seqkit
#
# To deactivate an active environment, use
#
#     $ conda deactivate




In [24]:
conda run -n seqkit seqkit stats --help

simple statistics of FASTA/Q files

Tips:
  1. For lots of small files (especially on SDD), use big value of '-j' to
     parallelize counting.

Usage:
  seqkit stats [flags]

Aliases:
  stats, stat

Flags:
  -a, --all                  all statistics, including quartiles of seq length, sum_gap, N50
  -b, --basename             only output basename of files
  -E, --fq-encoding string   fastq quality encoding. available values: 'sanger', 'solexa', 'illumina-1.3+', 'illumina-1.5+', 'illumina-1.8+'. (default "sanger")
  -G, --gap-letters string   gap letters (default "- .")
  -h, --help                 help for stats
  -e, --skip-err             skip error, only show warning message
  -i, --stdin-label string   label for replacing default "-" for stdin (default "-")
  -T, --tabular              output in machine-friendly tabular format

Global Flags:
      --alphabet-guess-seq-length int   length of sequence prefix of the first FASTA record based on which seqkit guesses the sequence type (

### 5.3.2. Print stats of contigs

To print basic stats on the assembled genome (`SRR3028792.fasta`) we can run the following:

In [25]:
conda run -n seqkit seqkit stats -a SRR3028792.fasta

file              format  type  num_seqs    sum_len  min_len  avg_len  max_len   Q1   Q2     Q3  sum_gap    N50  Q20(%)  Q30(%)
SRR3028792.fasta  FASTA   DNA      3,463  3,579,102      323  1,033.5   33,114  505  777  1,272        0  1,296       0       0




This gives us some basic information on our assembled genome. For example `sum_len` is the total length of all contigs (and so our genome, around 3.6 million bp, though keep in mind we are using a subset of all the sequence reads available so some data may be missing). `num_seqs` is the number of contigs in our genome.

# 6. Evaulate assembly quality (`quast`)

While `seqkit stats` gives us a quick way to examine the assembled genome, it can also be nice to have a visual and interactive report. This is where [quast](https://github.com/ablab/quast) comes in, as it can produce an interactive report that you can open in your web browser.

You can upload the `SRR3028792.fasta` file to the quast website at <http://cab.cc.spbu.ru/quast/>. However, we will go over how to run quast on the command-line (useful if you have many genomes you want to examine at once).

## 6.1. Install `quast`

The first step is to install a local version of `quast` using conda.

In [26]:
conda create -y -n quast quast

Solving environment: done

## Package Plan ##

  environment location: /home/CSCScience.ca/apetkau/miniconda3/envs/quast

  added / updated specs:
    - quast


The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu
  blast              bioconda/linux-64::blast-2.12.0-pl5262h3289130_0
  brotli             conda-forge/linux-64::brotli-1.0.9-h7f98852_6
  brotli-bin         conda-forge/linux-64::brotli-bin-1.0.9-h7f98852_6
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0
  ca-certificates    conda-forge/linux-64::ca-certificates-2021.10.8-ha878542_0
  certifi            conda-forge/linux-64::certifi-2021.10.8-py37h89c1867_1
  circos             bioconda/noarch::circos-0.69.8-hdfd78af_1
  curl               conda-forge/linux-64::curl-7.81.0-h494985f_0
  cycler             cond

In [27]:
conda run -n quast quast --help

QUAST: Quality Assessment Tool for Genome Assemblies
Version: 5.0.2

Usage: python /home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast [options] <files_with_contigs>

Options:
-o  --output-dir  <dirname>       Directory to store all result files [default: quast_results/results_<datetime>]
-r                <filename>      Reference genome file
-g  --features [type:]<filename>  File with genomic feature coordinates in the reference (GFF, BED, NCBI or TXT)
                                  Optional 'type' can be specified for extracting only a specific feature type from GFF
-m  --min-contig  <int>           Lower threshold for contig length [default: 500]
-t  --threads     <int>           Maximum number of threads [default: 25% of CPUs]

Advanced options:
-s  --split-scaffolds                 Split assemblies by continuous fragments of N's and add such "contigs" to the comparison
-l  --labels "label, label, ..."      Names of assemblies to use in reports, comma-separated. If cont

## 6.2. Create a quality report with `quast`

Now we can run quast as follows:

In [28]:
conda run -n quast quast SRR3028792.fasta

/home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast SRR3028792.fasta

Version: 5.0.2

System information:
  OS: Linux-4.15.0-161-generic-x86_64-with-debian-buster-sid (linux_64)
  Python version: 3.7.12
  CPUs number: 56

Started: 2022-01-24 10:26:18

Logging to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_24_10_26_18/quast.log
NOTICE: Maximum number of threads is set to 14 (use --threads option to set it manually)

CWD: /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial
Main parameters: 
  MODE: default, threads: 14, minimum contig length: 500, minimum alignment length: 65, \
  ambiguity: one, threshold for extensive misassembly size: 1000

Contigs:
  Pre-processing...
  SRR3028792.fasta ==> SRR3028792

2022-01-24 10:26:18
Running Basic statistics processor...
  Contig files: 
    SRR3028792
  Calculating N50 and L50...
    SRR

If you have more than one genome to evaluate you can run list them all on the command line `quast genome1.fasta genome2.fasta ...` or just use `quast *.fasta`.

### 6.2.1 Zip up report for download

The output of quast can be found in the directory `quast_results/latest/report.html` and you can open this in a web browser. However, if you are running this using Jupyter then we have to download the entire directory to your local computer first before viewing.

We can download the entire `quast_results/` directory by first zipping up the quast report. You can create a `*.zip` archive from the command-line with the `zip` command.

In [29]:
zip -r quast.zip quast_results

  adding: quast_results/ (stored 0%)
  adding: quast_results/latest/ (stored 0%)
  adding: quast_results/latest/report.tsv (deflated 56%)
  adding: quast_results/latest/quast.log (deflated 71%)
  adding: quast_results/latest/transposed_report.tsv (deflated 57%)
  adding: quast_results/latest/report.pdf (deflated 25%)
  adding: quast_results/latest/icarus.html (deflated 85%)
  adding: quast_results/latest/report.tex (deflated 64%)
  adding: quast_results/latest/transposed_report.txt (deflated 67%)
  adding: quast_results/latest/transposed_report.tex (deflated 59%)
  adding: quast_results/latest/report.html (deflated 72%)
  adding: quast_results/latest/basic_stats/ (stored 0%)
  adding: quast_results/latest/basic_stats/SRR3028792_GC_content_plot.pdf (deflated 29%)
  adding: quast_results/latest/basic_stats/cumulative_plot.pdf (deflated 30%)
  adding: quast_results/latest/basic_stats/GC_content_plot.pdf (deflated 28%)
  adding: quast_results/latest/basic_stats/Nx_plot.pdf (deflated 26%)
 

Now you can download the zip file by right-clicking `quast.zip` and selecting **Download**.

![jupyter-quast-zip.png](../images/jupyter-quast-zip.png)

Next, you can unzip the `quast.zip` file on your local computer and open `quast_report/latest/report.html` in a web browser. ou should end up with something that looks like:

![quast-report.png](../images/quast-report.png)

# 7. Compare our assembly to an assembly with all the data

An assemble of this same genome using all the sequence reads available can be found at `example/SRR3028792-full.fasta`. In order to get an idea of how to use quast to compare the quality of different assembled genomes we will generate a new report with both these genomes.

To do this we can re-run quast and pass both genomes as arguments.

In [30]:
conda run -n quast quast SRR3028792.fasta example/SRR3028792-full.fasta

/home/CSCScience.ca/apetkau/miniconda3/envs/quast/bin/quast SRR3028792.fasta example/SRR3028792-full.fasta

Version: 5.0.2

System information:
  OS: Linux-4.15.0-161-generic-x86_64-with-debian-buster-sid (linux_64)
  Python version: 3.7.12
  CPUs number: 56

Started: 2022-01-24 10:26:21

Logging to /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/quast_results/results_2022_01_24_10_26_21/quast.log
NOTICE: Maximum number of threads is set to 14 (use --threads option to set it manually)

CWD: /home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial
Main parameters: 
  MODE: default, threads: 14, minimum contig length: 500, minimum alignment length: 65, \
  ambiguity: one, threshold for extensive misassembly size: 1000

Contigs:
  Pre-processing...
  1  SRR3028792.fasta ==> SRR3028792
  2  example/SRR3028792-full.fasta ==> SRR3028792_full

2022-01-24 10:26:22
Running Basic stat

Now let's zip up the quast results again.

In [31]:
zip -r quast.zip quast_results

updating: quast_results/ (stored 0%)
updating: quast_results/latest/ (stored 0%)
updating: quast_results/latest/report.tsv (deflated 55%)
updating: quast_results/latest/quast.log (deflated 73%)
updating: quast_results/latest/transposed_report.tsv (deflated 56%)
updating: quast_results/latest/report.pdf (deflated 25%)
updating: quast_results/latest/icarus.html (deflated 85%)
updating: quast_results/latest/report.tex (deflated 64%)
updating: quast_results/latest/transposed_report.txt (deflated 70%)
updating: quast_results/latest/transposed_report.tex (deflated 59%)
updating: quast_results/latest/report.html (deflated 72%)
updating: quast_results/latest/basic_stats/ (stored 0%)
updating: quast_results/latest/basic_stats/SRR3028792_GC_content_plot.pdf (deflated 29%)
updating: quast_results/latest/basic_stats/cumulative_plot.pdf (deflated 29%)
updating: quast_results/latest/basic_stats/GC_content_plot.pdf (deflated 28%)
updating: quast_results/latest/basic_stats/Nx_plot.pdf (deflated 30%)
u

Now re-download `quast.zip` and open up `quast_results/latest/report.html` in your browser. You should see something like the following:

![quast-both-report.png](../images/quast-both-report.png)


Try comparing some of this information between both genomes (**Total Length**, **N50**, etc). Which assembly is better?

*Note: You can find a pre-computed set of QUAST results at `example/quast.zip`.*

# 8. Celebrate

Hooray, you are finished 😄🥳❤️💻. Congratulations. For additional ways to use the genome assembly you have just generated please see the next steps.

# 9. Where to go from here

An assembled genome file (`SRR3028792.fasta`) often forms the basis for other downstream analysis. Here are a few examples you can try on your own.

## 9.1. Anti-microbial resistance

Try uploading `SRR3028792.fasta` to CARD/RGI (which will identify genes associated with antimicrobial resistance): https://card.mcmaster.ca/analyze/rgi

## 9.2. Multi-locus sequence typing

Try downloading and running [mlst](https://github.com/tseemann/mlst) on the `SRR3028792.fasta` file, which will give you the [Multi-locus sequence type](https://en.wikipedia.org/wiki/Multilocus_sequence_typing) of the genome. You can do this with the following commands.

First we will install [mamba](https://mamba.readthedocs.io/en/latest/), which is a faster version of `conda`. For most cases using the `conda` command is okay, but sometimes conda is slow at installing tools and `mamba` is meant to speed things up. I found that `conda` was too slow to install `mlst`.

In [32]:
conda install -y mamba

Solving environment: done

# All requested packages already installed.

(base) 


Next we can use `mamba` to install `mlst`. `mamba` is meant to replace `conda` so you can run the same `conda create` command as you're used to be replace it with `mamba create`.

In [33]:
# Text starting with "#" is a comment as not a command
# The below is the equivalent `conda` command which we are replacing with `mamba`
# I use "-q" to print less information when running mamba
# conda create -y -n mlst mlst
mamba create -q -y -n mlst mlst

(base) (base) (base) (base)   Package                                       Version  Build                 Channel                    Size
────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex                                   0.1  conda_forge           conda-forge/linux-64     Cached
  + _openmp_mutex                                   4.5  1_gnu                 conda-forge/linux-64     Cached
  + any2fasta                                     0.4.2  hdfd78af_3            bioconda/noarch          Cached
  + blast                                        2.12.0  pl5262h3289130_0      bioconda/linux-64        Cached
  + bzip2                                         1.0.8  h7f98852_4            conda-forge/linux-64     Cached
  + c-ares                                       1.18.1  h7f98852_0 

Once the tools are installed with `mamba` we used the regular `conda` command to run.

In [34]:
conda run -n mlst mlst SRR3028792.fasta

SRR3028792.fasta	senterica	-	aroC(2)	dnaN(1034?)	hemD(828?)	hisD(1495?)	purE(1129?)	sucA(9)	thrA(1089?)

[10:27:04] This is mlst 2.19.0 running on linux with Perl 5.026002
[10:27:04] Checking mlst dependencies:
[10:27:04] Found 'blastn' => /home/CSCScience.ca/apetkau/miniconda3/envs/mlst/bin/blastn
[10:27:04] Found 'any2fasta' => /home/CSCScience.ca/apetkau/miniconda3/envs/mlst/bin/any2fasta
[10:27:05] Found blastn: 2.12.0+ (002012)
[10:27:05] Excluding 2 schemes: abaumannii ecoli_2
[10:27:07] Found exact allele match senterica.aroC-2
[10:27:07] Found exact allele match senterica.sucA-9
[10:27:07] If you have problems, please file at https://github.com/tseemann/mlst/issues
[10:27:07] Done.

(base) 


## 9.3. Serotyping

As this is a *Salmonella* genome, try running the [sistr](https://github.com/phac-nml/sistr_cmd) command on the genome to get the [serotype](https://en.wikipedia.org/wiki/Serotype).

In [35]:
mamba create -q -y -n sistr sistr_cmd

  Package                         Version  Build                Channel                    Size
─────────────────────────────────────────────────────────────────────────────────────────────────
  Install:
─────────────────────────────────────────────────────────────────────────────────────────────────

  + _libgcc_mutex                     0.1  conda_forge          conda-forge/linux-64     Cached
  + _openmp_mutex                     4.5  1_gnu                conda-forge/linux-64     Cached
  + blast                          2.12.0  pl5262h3289130_0     bioconda/linux-64        Cached
  + blosc                          1.21.0  h9c3ff4c_0           conda-forge/linux-64     Cached
  + bzip2                           1.0.8  h7f98852_4           conda-forge/linux-64     Cached
  + c-ares                         1.18.1  h7f98852_0           conda-forge/linux-64     Cached
  + ca-certificates             2021.10.8  ha878542_0           conda-forge/linux-64     Cached
  + capnproto           

In [36]:
conda run -n sistr sistr --help

usage: sistr_cmd [-h] [-i fasta_path genome_name] [-f OUTPUT_FORMAT]
                 [-o OUTPUT_PREDICTION] [-M] [-p CGMLST_PROFILES]
                 [-n NOVEL_ALLELES] [-a ALLELES_OUTPUT] [-T TMP_DIR] [-K]
                 [--use-full-cgmlst-db] [--no-cgmlst] [-m] [--qc] [-t THREADS]
                 [-v] [-V]
                 [F [F ...]]

SISTR (Salmonella In Silico Typing Resource) Command-line Tool
Serovar predictions from whole-genome sequence assemblies by determination of antigen gene and cgMLST gene alleles using BLAST.

Note about using the "--use-full-cgmlst-db" flag:
    The "centroid" allele database is ~10% the size of the full set so analysis is much quicker with the "centroid" vs "full" set of alleles. Results between 2 cgMLST allele sets should not differ.

If you find this program useful in your research, please cite as:

The Salmonella In Silico Typing Resource (SISTR): an open web-accessible tool for rapidly typing and subtyping draft Salmonella genome assemblies.


In [37]:
conda run -n sistr sistr -f csv -o SRR3028792.sistr.csv SRR3028792.fasta

2022-01-24 10:27:28,188 ERROR: Too many gapped sites in extracted allele seq for marker NZ_AOXE01000068.1_72 contained 414 gaps out of 780 bp (0.5307692307692308 > 0.05); stitle: Contig_76_6.00195 [in /home/CSCScience.ca/apetkau/miniconda3/envs/sistr/lib/python3.8/site-packages/sistr/src/cgmlst/__init__.py:162]
2022-01-24 10:27:28,472 ERROR: Too many gapped sites in extracted allele seq for marker NZ_AOXE01000034.1_53 contained 645 gaps out of 1404 bp (0.4594017094017094 > 0.05); stitle: Contig_88_8.21703 [in /home/CSCScience.ca/apetkau/miniconda3/envs/sistr/lib/python3.8/site-packages/sistr/src/cgmlst/__init__.py:162]
2022-01-24 10:27:28,742 ERROR: Too many gapped sites in extracted allele seq for marker NZ_AOXE01000003.1_57 contained 79 gaps out of 174 bp (0.4540229885057471 > 0.05); stitle: Contig_146_7.13532 [in /home/CSCScience.ca/apetkau/miniconda3/envs/sistr/lib/python3.8/site-packages/sistr/src/cgmlst/__init__.py:162]
2022-01-24 10:27:29,025 ERROR: Too many gapped sites in extr

The output is saved to the file `SRR3028792.sistr.csv`. We can view it with `cat`. Also, since it's a **csv** file, you can download it and open it in any spreadsheet software (e.g., Excel).

*Note: messages you see like `ERROR: Missing cgmlst_results for NZ_AOXE01000052.1_131` are because we ran this on an assembled genome using only **10%** of the original data, so some genes on the genome are missing from the assembly.*

In [38]:
cat SRR3028792.sistr.csv

cgmlst_ST,cgmlst_distance,cgmlst_found_loci,cgmlst_genome_match,cgmlst_matching_alleles,cgmlst_subspecies,fasta_filepath,genome,h1,h2,o_antigen,serogroup,serovar,serovar_antigen,serovar_cgmlst
,0.5878787878787879,148,2017-MER-0343,136,enterica,/home/CSCScience.ca/apetkau/workspace/2022-01-26-Downloading-and-assembling-microbial-sequence-data/tutorial/SRR3028792.fasta,SRR3028792,r,-,"1,4,[5],12",B,"I 1,4,[5],12:r:-","I 1,4,[5],12:r:-",Heidelberg
(base) 
